### PR TITLE
Check APT Package Versions

### PR DESCRIPTION
**Pull request by Natrinicle**
Make sure to keep the cache up to date as package versions change
upstream. Also, allow specific package versions to be selected.

### CODE REVIEW

**Comment by Natrinicle** [comment]
When running with

```yaml
      - uses: Natrinicle/cache-apt-pkgs-action@check-package-versions
        with:
          packages: wkhtmltopdf xvfb
          version: 1.0
```

The output is
```
Run Natrinicle/cache-apt-pkgs-action@check-package-versions
  with:
    packages: wkhtmltopdf xvfb
    version: 1
    refresh: false
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.9.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.10/x64/lib
Run /home/runner/work/_actions/Natrinicle/cache-apt-pkgs-action/check-package-versions/pre_cache_action.sh \
  /home/runner/work/_actions/Natrinicle/cache-apt-pkgs-action/check-package-versions/pre_cache_action.sh \
    ~/cache-apt-pkgs \
    "1" \
    wkhtmltopdf xvfb
  echo "CACHE_KEY=$(cat ~/cache-apt-pkgs/cache_key.md5)" >> $GITHUB_ENV
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.9.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.10/x64/lib
Validating action arguments (version='1', packages='wkhtmltopdf xvfb ')...done.
Updating APT package list...done.

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.


WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Verifying packages...done.
Creating cache key...
- Normalized package list is 'wkhtmltopdf=0.12.5-1build1 xvfb=0.12.5-1build1'.
- Value to hash is 'wkhtmltopdf=0.12.5-1build1 xvfb=0.12.5-1build1 @ 1'.
- Value hashed as 'ced7b66c9d6d90c7d33a6be474ad3de4'.
done.
Hash value written to /home/runner/cache-apt-pkgs/cache_key.md5
Run actions/cache@v2
  with:
    path: ~/cache-apt-pkgs
    key: cache-apt-pkgs_ced7b66c9d6d90c7d33a6be474ad3de4
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.9.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.10/x64/lib
    CACHE_KEY: ced7b66c9d6d90c7d33a6be474ad3de4
Received 1334376 of 1334376 (100.0%), 13.8 MBs/sec
Cache Size: ~1 MB (1334376 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/8b793ca3-5e84-4da2-9c73-2bbb9f1f053c/cache.tzst -P -C /home/runner/work/billboard-server/billboard-server
Cache restored successfully

Cache restored from key: cache-apt-pkgs_ced7b66c9d6d90c7d33a6be474ad3de4
Run /home/runner/work/_actions/Natrinicle/cache-apt-pkgs-action/check-package-versions/post_cache_action.sh \
  /home/runner/work/_actions/Natrinicle/cache-apt-pkgs-action/check-package-versions/post_cache_action.sh \
    ~/cache-apt-pkgs \
    / \
    "true" \
    wkhtmltopdf xvfb
  echo "::set-output name=package-version-list::$(cat ~/cache-apt-pkgs/manifest.log)"
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.9.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.10/x64/lib
    CACHE_KEY: ced7b66c9d6d90c7d33a6be474ad3de4
Found 4 files in the cache.
- cache_key.md5
- manifest.log
- wkhtmltopdf.tar.gz
- xvfb.tar.gz
Reading from manifest...
- wkhtmltopdf 0.12.5-1build1
- xvfb 2 1.20.13-1ubuntu1~20.04.2
done.
Restoring 2 packages from cache...
- wkhtmltopdf.tar.gz restoring...done.
- xvfb.tar.gz restoring...done.
done.
```

File: install_and_cache_pkgs.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ for package in $packages; do
 
   echo -n "  Caching to $cache_filepath..."
```
Comment (awalsh128) [inline install_and_cache_pkgs.sh:32]: Thanks for updating the variable references to a safe syntax. Could you standardize these variables as well?

File: install_and_cache_pkgs.sh [hunk 2]
```diff
@@ -31,7 +31,7 @@ for package in $packages; do
...
   # Pipe all package files (no folders) to Tar.
-  dpkg -L $package |
+  dpkg -L "$(echo "${package}" | cut -d"=" -f1)" |
```
Comment (awalsh128) [inline install_and_cache_pkgs.sh:34]: Create a new variable that is self descriptive or add a comment please.

File: pre_cache_action.sh [hunk 3]
```diff
@@ -1,40 +1,56 @@
...
+echo "done."
+
 echo -n "Verifying packages..."
-for package in $packages; do
-  escaped=$(echo $package | sed 's/+/\\+/g')
-  apt-cache search ^$escaped$ | grep $package > /dev/null
-  if [ $? -ne 0 ]; then
+for package in ${packages}; do
+  if echo "${package}" | grep -q "="; then
+    pkg_name=$(echo "${package}" | cut -d "=" -f1)
```
Comment (awalsh128) [inline pre_cache_action.sh:39]: package_name and package_ver to keep it consistent.

File: pre_cache_action.sh [hunk 4]
```diff
@@ -43,18 +59,18 @@ set -e
 
 echo "Creating cache key..."
 
-# Remove package delimiters, sort (requires newline) and then convert back to inline list.
-normalized_list=$(echo $packages | sed 's/[\s,]+/ /g' | tr ' ' '\n' | sort | tr '\n' ' ')
-echo "- Normalized package list is '$normalized_list'."
+# Remove extraneous spaces
+package_list="$(echo "${package_list}" | sed 's/\s\+/ /g;s/^\s\+//g;s/\s\+$//g')"
```
Comment (awalsh128) [inline pre_cache_action.sh:63]: Thanks, much better.

**Review by awalsh128 — CHANGES_REQUESTED** [review]
Thanks for creating these changes. If you want to proceed I can pull and update myself. If you'd rather just get it pulled as-is, could you create a regression test as well? The CI is a separate project under https://github.com/awalsh128/cache-apt-pkgs-action-ci

**Comment by Natrinicle** [comment]
> Thanks for creating these changes. If you want to proceed I can pull and update myself. If you'd rather just get it pulled as-is, could you create a regression test as well? The CI is a separate project under https://github.com/awalsh128/cache-apt-pkgs-action-ci

First off, thanks for the in-depth and thoughtful review! It's one of the best I've seen.

I actually hit another issue that I was just beginning to fix in another branch. This other branch includes the safer syntax in the install_and_cache_pkgs.sh file but also reads the output of an apt-get --dry-run to look for all installed packages. I found when trying to install wkhtmltopdf it required something like 15 other packages. When the cache was restored, it was missing key files necessary for the binary to run. Instead of trying to manually list those packages in the workflow and needing to keep that up to date as package requirements changed upstream, I thought it would be better to just programmatically cache those as well.

I can either fix this branch with the suggestions you made or take the changes you suggested and apply them to the other branch, kill this PR, and start a new one. The other branch comparison can be found here: https://github.com/Natrinicle/cache-apt-pkgs-action/compare/master...cache-non-explicitly-listed-pacakges

Let me know what you'd prefer and I'll go from there.

**Comment by awalsh128** [comment]
> I can either fix this branch with the suggestions you made or take the changes you suggested and apply them to the other branch, kill this PR, and start a new one. 

Let's just kill it and merge them. The rollback will sort of suck in terms of debugging but I think we should be fine since you have been testing it.

**Pull request closed without merging**